mavros/src/plugins/wind_estimation.cpp

```diff
@@ -0,0 +1,96 @@
+/**
+ * @brief wind estimation plugin
+ * @file wind_estimation.cpp
+ * @author Thomas Stastny <thomas.stastny@mavt.ethz.ch>
+ *
+ * @addtogroup plugin
+ * @{
+ */
+/*
+ * Copyright 2018 Thomas Stastny <thomas.stastny@mavt.ethz.ch>
+ *
+ * This file is part of the mavros package and subject to the license terms
+ * in the top-level LICENSE file of the mavros repository.
+ * https://github.com/mavlink/mavros/tree/master/LICENSE.md
+ */
+
+#include <angles/angles.h>
+#include <mavros/mavros_plugin.h>
+
+#include <geometry_msgs/TwistWithCovarianceStamped.h>
+
+namespace mavros {
+namespace std_plugins {
+/**
+ * @brief Wind estimation plugin.
+ */
+class WindEstimationPlugin : public plugin::PluginBase {
+public:
+	WindEstimationPlugin() : PluginBase(),
+		nh("~")
+	{ }
+
+	/**
+	 * Plugin initializer. Constructor should not do this.
+	 */
+	void initialize(UAS &uas_)
+	{
+		PluginBase::initialize(uas_);
+
+		wind_pub = nh.advertise<geometry_msgs::TwistWithCovarianceStamped>("wind_estimation", 10);
+	}
+
+	Subscriptions get_subscriptions()
+	{
+		return {
+			make_handler(&WindEstimationPlugin::handle_APM_wind),
+			make_handler(&WindEstimationPlugin::handle_PX4_wind),
+		};
+	}
+
+private:
+	ros::NodeHandle nh;
+
+	ros::Publisher wind_pub;
+
+	/**
+	 * Handle APM specific wind direction estimation message
+	 */
+	void handle_APM_wind(const mavlink::mavlink_message_t *msg, mavlink::ardupilotmega::msg::WIND &wind)
+	{
+		const double speed = wind.speed;
+		const double course = angles::from_degrees(wind.direction);
+
+		auto twist_cov = boost::make_shared<geometry_msgs::TwistWithCovarianceStamped>();
+		twist_cov->header.stamp = ros::Time::now();
+		// TODO: check math's
+		twist_cov->twist.twist.linear.x = speed * std::sin(course);
+		twist_cov->twist.twist.linear.y = speed * std::cos(course);
+		twist_cov->twist.twist.linear.z = wind.speed_z;
+
+		wind_pub.publish(twist_cov);
+	}
+
+	/**
+	 * Handle APM specific wind direction estimation message
+	 */
+	void handle_PX4_wind(const mavlink::mavlink_message_t *msg, mavlink::common::msg::WIND_COV &wind)
+	{
+		auto twist_cov = boost::make_shared<geometry_msgs::TwistWithCovarianceStamped>();
+		twist_cov->header.stamp = ros::Time::now();
+
+		twist_cov->twist.twist.linear.x = wind.wind_x;
+		twist_cov->twist.twist.linear.y = wind.wind_y;
+		twist_cov->twist.twist.linear.z = wind.wind_z;
+
+		twist_cov->twist.covariance[0] = wind.var_horiz; // NOTE: this is a summed covariance for both x and y horizontal wind components
```

Comment:
What about splitting this on the Firmware side also?

Reply:
Problem is in the mavlink message itself, (WIND_COV)[https://mavlink.io/en/messages/common.html#WIND_COV]. Only has "horizontal" and "vertical" as members. As I understand it, Mavlink is a "write only" process, correct? Otherwise, I'd for sure be in favor of splitting them.

Reply:
MAVLink can be extended. One does not have to replace fields. You can so update the description of the fields.

Reply:
I guess that would be a bit of an odd update though. "var_horiz" and "var_horiz2"? My guess would be that they originally put only the summed covariance for horizontal as they weren't expecting the horizontal components to have that much different of an uncertainty anyway (a somewhat reasonable expectation looking at how the wind is estimated). Though for sure vertical and horizontal will have different uncertainties. Do you think it is worth to add a field at this point? Or just keep it as it is for now? If you want to change it, could you recommend a field name?

Reply:
If it is a sum, would it be reasonable to consider a split by half to the two components (X and Y)?

Reply:
So I checked a recent flight log -- looks like maybe not the best to assume they are equal (though they are somehow close). see
![wind_cov1](https://user-images.githubusercontent.com/8026163/39117374-160ccbcc-46e7-11e8-8113-f0632d464f56.png) and (zoomed)
![wind_cov2](https://user-images.githubusercontent.com/8026163/39117381-1a175e9e-46e7-11e8-8b36-b833891ef19a.png)

I would probably be for leaving it as is for the time being, and if someone wants more ekf detailed info -- maybe we would need to make a new mavlink msg specifically for that, leaving this message as a common default one for people who dont care to have every piece of info. Otherwise your other suggestion of extending somehow the current message with a few extra covariance values could also be an option.

I actually don't think vertical wind is even estimated on EKF2 currently. Though I may look into some model-based extensions to help with this later this summer. But for now PX4 populates that field = 0.